### PR TITLE
[REVIEW] Add List types support in data generator

### PR DESCRIPTION
**Pull request by galipremsagar**
Resolves: #6263 

This PR introduces changes which will enable generation of random list columns in datagenerator which will be used as part of fuzz tests.

cc: @vuule

### CODE REVIEW

**Comment by brandon-b-miller** [comment]
Looked over this mostly high-level so far, still getting into the details, but looks great so far. More to come.

File: python/cudf/cudf/_fuzz_testing/tests/fuzz_test_csv.py [hunk 1]
```diff
@@ -40,7 +40,7 @@ def csv_writer_test(pdf):
 @pythonfuzz(
     data_handle=CSVWriter,
     params={
-        "sep": list([",", "|", "\t", "\t\t", "~"]),
+        "sep": list([",", "|", "\t", "\r", "~"]),
```
Comment (brandon-b-miller) [inline python/cudf/cudf/_fuzz_testing/tests/fuzz_test_csv.py:43]: why this change?

**Comment by brandon-b-miller** [comment]
Implementation of the list generation process looks fine, the rest just looks to be plumbing and moving some code around which also looks fine. One question I had was have you tried to see if some of the combinations of input parameters can OOM the average GPU? It seems that there might be some combinations of `lists_max_length ` and `nesting_max_depth` that could result in large datasets and cause an error that isn't related to I/O.

File: python/cudf/cudf/_fuzz_testing/tests/fuzz_test_csv.py (hunk 1 above)
Comment (galipremsagar) [reply python/cudf/cudf/_fuzz_testing/tests/fuzz_test_csv.py:43]: Had to make this change because we don't support multi-characters in `sep`: https://github.com/rapidsai/cudf/pull/7095

**Comment by galipremsagar** [comment]
> Implementation of the list generation process looks fine, the rest just looks to be plumbing and moving some code around which also looks fine. One question I had was have you tried to see if some of the combinations of input parameters can OOM the average GPU? It seems that there might be some combinations of `lists_max_length ` and `nesting_max_depth` that could result in large datasets and cause an error that isn't related to I/O.


Yes, we can OOM easily with high values for those params, same is the case with strings too when `max_string_length` & `size` are too high we'll easily OOM an average GPU. In these cases where the data generation itself fails, this particular line will fail: https://github.com/rapidsai/cudf/blob/branch-0.18/python/cudf/cudf/_fuzz_testing/fuzzer.py#L86, which will not be recorded into fuzz failure logs and the entire fuzz worker stops.

File: python/cudf/cudf/tests/dataset_generator.py [hunk 2]
```diff
@@ -436,4 +489,86 @@ def timedelta_generator(dtype, size):
 
 
 def boolean_generator(size):
+    """
+    Generator for bool data
+    """
     return lambda: np.random.choice(a=[False, True], size=size)
+
+
+def get_nested_lists(dtype, nesting_depth, list_max_length):
```
Comment (vuule) [inline python/cudf/cudf/tests/dataset_generator.py:498]: Man, it's hard to be sure if this is correct.
Here's what I did in C++, might be simpler to do something similar here:
Generate `num_rows` * `list_len`^`depth` elements
`depth` times:
  Split the list into a list of lists, where the "cut" offsets are generated incrementally using a distribution around the list_len parameter. You need `num_rows` * `list_len`^(`current_depth` - 1) offsets in each level (or until offset is > current length).
If this sounds like it could be easy to implement in Python, I would suggest it. AFAICT, the length distribution would be better. Otherwise we can try this one and see if it finds us some bugs.

Comment (brandon-b-miller) [reply python/cudf/cudf/tests/dataset_generator.py:498]: I spent a couple hours looking at this and couldn't come up with anything that was obviously better. FWIW I did come up with a recursive approach that I think gets the right answer:

```
top_level_list = []
def make_lists(list_max_len, nesting_depth, top_level_list):
    nesting_depth -= 1
    if nesting_depth > 0:
        L = np.random.randint(1, list_max_len)
        for i in range(L):
            top_level_list.append(make_lists(list_max_len, nesting_depth, []))
    else:
        top_level_list = np.random.randint(0, list_max_len, np.random.randint(1, list_max_len))
    return top_level_list
```

Comment (vuule) [reply python/cudf/cudf/tests/dataset_generator.py:498]: This looks like a correct, straight-forward implementation. Thanks @brandon-b-miller !

Comment (galipremsagar) [reply python/cudf/cudf/tests/dataset_generator.py]: > This looks like a correct, straight-forward implementation. Thanks @brandon-b-miller !

+1

The recursive approach looks easier to maintain too, I plugged it into the existing helper functions and tested this locally, and verified for all different value types and works perfectly 👍

Comment (galipremsagar) [reply python/cudf/cudf/tests/dataset_generator.py]: @vuule @brandon-b-miller This is ready for another round of review.

**Review by vuule — APPROVED** [review]

**Review by brandon-b-miller — APPROVED** [review]

**Comment by galipremsagar** [comment]
@gpucibot merge
